File: src/hcmai/api/contracts/feedback.py
"""Contracts for KIS chat feedback, turn requests, and state responses."""

from __future__ import annotations

from typing import Annotated, Any, Literal, Self
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, model_validator

from hcmai.api.contracts.event_trail import EventTrailStateResponse
from hcmai.api.contracts.kis import KISSearchResult
from hcmai.kis.models import EventId, KISIntent

NonBlank = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


class RetrievalOverride(BaseModel):
    """Retrieval text overrides for dense and/or lexical retrieval for one event."""

    model_config = ConfigDict(extra="forbid")
    dense_text: str | None = None
    bm25_text: str | None = None


class FeedbackOpenRequest(BaseModel):
    """Payload to open a bounded feedback session from an existing search result."""

    model_config = ConfigDict(extra="forbid")

    intent: KISIntent
    original_query: NonBlank
    evidence_snapshot_id: NonBlank
    use_dense: bool = True
    use_bm25: bool = True
    top_k: int = Field(default=20, ge=1)
    initial_results: list[KISSearchResult] = Field(default_factory=list)


class FeedbackTurnRequest(BaseModel):
    """One feedback interaction turn initiated by user chat or shortcut button."""

    model_config = ConfigDict(extra="forbid")

    request_id: NonBlank
    expected_feedback_revision: int = Field(ge=0)
    expected_kis_revision: int = Field(ge=0)
    expected_trail_revision: int | None = None
    message: NonBlank
    selected_result_id: str | None = None
    selected_event_id: str | None = None
    selected_frame_id: str | None = None


class FeedbackUndoRequest(BaseModel):
    """Request to revert the most recent committed feedback turn."""

    model_config = ConfigDict(extra="forbid")

    request_id: NonBlank
    expected_feedback_revision: int = Field(ge=0)


class FeedbackStateResponse(BaseModel):
    """Complete state envelope returned by feedback open, turn, and undo operations."""

    model_config = ConfigDict(extra="forbid")

    session_id: str
    status: Literal["applied", "clarification", "exhausted", "proposal"]
    feedback_revision: int = Field(ge=1)
    intent: KISIntent
    retrieval_overrides: dict[str, RetrievalOverride] = Field(default_factory=dict)
    results: list[KISSearchResult] = Field(default_factory=list)
    evidence_snapshot_id: str
    trail: EventTrailStateResponse | None = None
    assistant_message: str | None = None
    query_proposal: dict[str, Any] | None = None
    changed_event_ids: list[str] = Field(default_factory=list)
    scope: str = "all_videos"
    can_undo: bool = False
    latency: dict[str, Any] | None = None


class FeedbackOpenResponse(BaseModel):
    """Response returned when opening a feedback session."""

    model_config = ConfigDict(extra="forbid")

    session_id: str
    feedback_revision: int
    state: FeedbackStateResponse


__all__ = [
    "FeedbackOpenRequest",
    "FeedbackOpenResponse",
    "FeedbackStateResponse",
    "FeedbackTurnRequest",
    "FeedbackUndoRequest",
    "RetrievalOverride",
]

File: frontend/src/api/feedback.js
/**
 * Thin HTTP client for KIS chat feedback, turn execution, and Undo.
 */

import { requestJson } from './client';

export const openFeedbackSession = async ({
  intent,
  originalQuery,
  evidenceSnapshotId,
  useDense = true,
  useBm25 = true,
  topK = 20,
  initialResults = [],
  signal,
}) => {
  if (!intent || typeof intent !== 'object') {
    throw new Error('Intent is required to open a feedback session');
  }
  if (!originalQuery || typeof originalQuery !== 'string') {
    throw new Error('Original query is required to open a feedback session');
  }
  if (!evidenceSnapshotId || typeof evidenceSnapshotId !== 'string') {
    throw new Error('Evidence snapshot ID is required to open a feedback session');
  }

  const body = {
    intent,
    original_query: originalQuery.trim(),
    evidence_snapshot_id: evidenceSnapshotId,
    use_dense: useDense,
    use_bm25: useBm25,
    top_k: topK,
    initial_results: Array.isArray(initialResults) ? initialResults : [],
  };

  const data = await requestJson('/api/v1/kis/feedback/open', {
    method: 'POST',
    body,
    signal,
  });

  if (!data?.session_id || typeof data?.feedback_revision !== 'number' || !data?.state) {
    throw new Error('Malformed feedback open response');
  }
  return data;
};

export const sendFeedbackTurn = async (
  sessionId,
  {
    requestId,
    expectedFeedbackRevision,
    expectedKisRevision,
    expectedTrailRevision,
    message,
    selectedResultId,
    selectedEventId,
    selectedFrameId,
  },
  { signal } = {},
) => {
  if (!sessionId) {
    throw new Error('Session ID is required to send feedback turn');
  }
  if (!requestId) {
    throw new Error('Request ID is required to send feedback turn');
  }
  if (!message || !message.trim()) {
    throw new Error('Message is required to send feedback turn');
  }

  const body = {
    request_id: requestId,
    expected_feedback_revision: expectedFeedbackRevision,
    expected_kis_revision: expectedKisRevision,
    message: message.trim(),
  };

  if (expectedTrailRevision !== undefined && expectedTrailRevision !== null) {
    body.expected_trail_revision = expectedTrailRevision;
  }
  if (selectedResultId) body.selected_result_id = selectedResultId;
  if (selectedEventId) body.selected_event_id = selectedEventId;
  if (selectedFrameId) body.selected_frame_id = selectedFrameId;

  const data = await requestJson(`/api/v1/kis/feedback/${encodeURIComponent(sessionId)}/turn`, {
    method: 'POST',
    body,
    signal,
  });

  if (!data?.session_id || typeof data?.feedback_revision !== 'number') {
    throw new Error('Malformed feedback turn response');
  }
  return data;
};

export const undoFeedback = async (
  sessionId,
  { requestId, expectedFeedbackRevision },
  { signal } = {},
) => {
  if (!sessionId) {
    throw new Error('Session ID is required to undo feedback');
  }
  if (!requestId) {
    throw new Error('Request ID is required to undo feedback');
  }

  const body = {
    request_id: requestId,
    expected_feedback_revision: expectedFeedbackRevision,
  };

  const data = await requestJson(`/api/v1/kis/feedback/${encodeURIComponent(sessionId)}/undo`, {
    method: 'POST',
    body,
    signal,
  });

  if (!data?.session_id || typeof data?.feedback_revision !== 'number') {
    throw new Error('Malformed feedback undo response');
  }
  return data;
};
